### ISSUE TITLE
[BUG] requires networking to be enabled during build

### ISSUE THREAD

**Issue by RoyalOughtness** [34N0/pam-authramp-rpm#1]
Sources for an rpm build should be pulled prior to build so that the network connection can be disabled during build for improved security.

An example of a rust app that does this is: https://src.fedoraproject.org/rpms/kata-containers/blob/rawhide/f/kata-containers.spec#_100

You can generate a spec like this using rust2rpm:
https://docs.fedoraproject.org/en-US/packaging-guidelines/Rust/

**Comment by RoyalOughtness** [comment]
also, FYI this is a requirement anyways if this package is going to target getting official fedora packaging.

**Comment by wandapeter** [comment]
Thank you for the input! :+1:

**Comment by 34N0** [comment]
I steered away from rust2rpm so far because i really dislike python. but i guess i will try it :smile_cat:

**Comment by wandapeter** [comment]
I tried it but the problem seems that some dependncies are not part of fedora? I will try further. dnf error from copr:
```
Problem 1: nothing provides requested (crate(cli-xtask/default) >= 0.8.0 with crate(cli-xtask/default) < 0.9.0~)
 Problem 2: nothing provides requested (crate(cli-xtask/lib-crate) >= 0.8.0 with crate(cli-xtask/lib-crate) < 0.9.0~)
 Problem 3: nothing provides requested (crate(cli-xtask/main) >= 0.8.0 with crate(cli-xtask/main) < 0.9.0~)
 Problem 4: nothing provides requested (crate(pam-bindings/default) >= 0.1.1 with crate(pam-bindings/default) < 0.2.0~)
 Problem 5: nothing provides requested (crate(pam-client/default) >= 0.5.0 with crate(pam-client/default) < 0.6.0~)
 Problem 6: nothing provides requested (crate(sysinfo/default) >= 0.30.0 with crate(sysinfo/default) < 0.31.0~)
 Problem 7: nothing provides requested (crate(xshell/default) >= 0.2.5 with crate(xshell/default) < 0.3.0~)```

**Comment by RoyalOughtness** [comment]
I think you need to define those as `BuildRequires` so it can pull from crates:

https://src.fedoraproject.org/rpms/kata-containers/blob/rawhide/f/kata-containers.spec#_100

**Comment by wandapeter** [comment]
newer versions of rust2rpm use the `%cargo_generate_buildrequires -a` to generate `BuildRequires` entries. Those problems are the entries which fail at build time.

**Comment by wandapeter** [comment]
i rewrote the tests in C and implemented the pam library locally. makes some things easier and even adds some features since i only support linux and the library is built for macos as well. those changes actually remove all missing dependencies and using `BuildRequires` should work now. Will test this later.

**Comment by RoyalOughtness** [comment]
Oh okay that works. I think all that was needed though was to pull in the deps before build time. Networking is enabled in the sources step.

**Comment by wandapeter** [comment]
> ```
> Problem 1: nothing provides requested (crate(cli-xtask/default) >= 0.8.0 with crate(cli-xtask/default) < 0.9.0~)
>  Problem 2: nothing provides requested (crate(cli-xtask/lib-crate) >= 0.8.0 with crate(cli-xtask/lib-crate) < 0.9.0~)
>  Problem 3: nothing provides requested (crate(cli-xtask/main) >= 0.8.0 with crate(cli-xtask/main) < 0.9.0~)
>  Problem 4: nothing provides requested (crate(pam-bindings/default) >= 0.1.1 with crate(pam-bindings/default) < 0.2.0~)
>  Problem 5: nothing provides requested (crate(pam-client/default) >= 0.5.0 with crate(pam-client/default) < 0.6.0~)
>  Problem 6: nothing provides requested (crate(sysinfo/default) >= 0.30.0 with crate(sysinfo/default) < 0.31.0~)
>  Problem 7: nothing provides requested (crate(xshell/default) >= 0.2.5 with crate(xshell/default) < 0.3.0~)```

Thosw crates weren't in the registries so pulling them before build and not by cargo wasn't possible. I ment those.

**Comment by RoyalOughtness** [comment]
Ah I see. How do those get pulled in currently? Is there another registry that needs to be added?

**Comment by wandapeter** [comment]
> Ah I see. How do those get pulled in currently? Is there another registry that needs to be added?

I do not understand it fully yet. I am guessing since fedora requires a review for packages in the official repository, all its dependencies have to be reviewed as well. So all cargo crates have to have their own reviews .rpm.

For instance the `log` crate is published in the official repository as ` rust-log-devel` and can even be installed via dnf. The consequence of this is that to publish a rust app in fedora all its dependencies have to be published first as well.

Also it seemed to require my dev dependencies as well which doesn't make sense and might be a bug. Anyways removing the failing dependencies made sense anyways. fixed with #3
